In [1]:
import re

import numpy as np
import pandas as pd

In [2]:
main_df = pd.read_csv('data/dst-3.0_16_1_hh_database.csv', sep=';')

----

### Задание 2.1 ###

Прочитайте данные с помощью библиотеки Pandas. Перед чтением обратите внимание на разделитель внутри файла.
Чему равна размерность таблицы? Введите её в виде кортежа (число строк, число столбцов).

In [3]:
main_df.shape

(44744, 12)

----

### Задание 2.2 ###

Выведите несколько первых (последних) строк таблицы, чтобы убедиться в том, что ваши данные не повреждены. Ознакомьтесь с признаками и их структурой.
В каком формате представлен порядок слов в признаке «Пол, возраст»?

***Ответ - <Пол , возраст , дата рождения>***

In [4]:
main_df.head(3)

Пол, возраст          ЗП  \
0    Мужчина ,  39 лет , родился 27 ноября 1979  29000 руб.   
1     Мужчина ,  60 лет , родился 20 марта 1959  40000 руб.   
2  Женщина ,  36 лет , родилась 12 августа 1982  20000 руб.   

  Ищет работу на должность:  \
0   Системный администратор   
1      Технический писатель   
2                  Оператор   

                        Город, переезд, командировки  \
0  Советск (Калининградская область) , не готов к...   
1  Королев , не готов к переезду , готов к редким...   
2  Тверь , не готова к переезду , не готова к ком...   

                                           Занятость  \
0  частичная занятость, проектная работа, полная ...   
1  частичная занятость, проектная работа, полная ...   
2                                   полная занятость   

                                              График  \
0  гибкий график, полный день, сменный график, ва...   
1  гибкий график, полный день, сменный график, уд...   
2                                        полный день   

                                         Опыт работы  \
0  Опыт работы 16 лет 10 месяцев  Август 2010 — п...   
1  Опыт работы 19 лет 5 месяцев  Январь 2000 — по...   
2  Опыт работы 10 лет 3 месяца  Октябрь 2004 — Де...   

  Последнее/нынешнее место работы  \
0         МАОУ "СОШ № 1 г.Немана"   
1    Временный трудовой коллектив   
2                    ПАО Сбербанк   

                       Последняя/нынешняя должность  \
0                           Системный администратор   
1  Менеджер проекта, Аналитик, Технический писатель   
2                               Кассир-операционист   

                                   Образование и ВУЗ Обновление резюме  \
0  Неоконченное высшее образование 2000  Балтийск...  16.04.2019 15:59   
1  Высшее образование 1981  Военно-космическая ак...  12.04.2019 08:42   
2  Среднее специальное образование 2002 Профессио...  16.04.2019 08:35   

                             Авто  
0  Имеется собственный автомобиль  
1                      Не указано  
2                      Не указано

----

### Задание 2.3 ###

Выведите основную информацию о числе непустых значений в столбцах и их типах в таблице.
Какой тип данных имеют все столбцы? Введите его в поле для ответа.

***Ответ - object***

In [5]:
main_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 12 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
dtypes: object(12)
memory usage: 4.1+

----

### Задание 2.4 ###

Обратите внимание на информацию о числе непустых значений.
Выберите все признаки, в которых есть пропуски в данных:

***Ответ***
- Последняя/нынешняя должность
- Опыт работы
- Последнее/нынешнее место работы

----

### Задание 2.5 ###

Выведите основную статистическую информацию о столбцах.
Сколько уникальных значений содержится в столбце «Опыт работы»?

***Ответ - 44413***

Какая из искомых должностей самая распространённая?

***Ответ - Системный администратор***

In [6]:
main_df.describe(include=['object'])

Пол, возраст          ЗП  \
count                                           44744       44744   
unique                                          16003         690   
top     Мужчина ,  32 года , родился 17 сентября 1986  50000 руб.   
freq                                               18        4064   

       Ищет работу на должность:  \
count                      44744   
unique                     14929   
top      Системный администратор   
freq                        3099   

                             Город, переезд, командировки         Занятость  \
count                                               44744             44744   
unique                                              10063                38   
top     Москва , не готов к переезду , не готов к кома...  полная занятость   
freq                                                 1261             30026   

             График                                        Опыт работы  \
count         44744                                              44576   
unique           47                                              44413   
top     полный день  Опыт работы 10 лет 8 месяцев  Апрель 2018 — по...   
freq          22727                                                  3   

                          Последнее/нынешнее место работы  \
count                                               44743   
unique                                              30214   
top     Индивидуальное предпринимательство / частная п...   
freq                                                  935   

       Последняя/нынешняя должность  \
count                         44742   
unique                        16927   
top         Системный администратор   
freq                           2062   

                                        Образование и ВУЗ Обновление резюме  \
count                                               44744             44744   
unique                                              40148             18838   
top     Высшее образование 1987 Военный инженерный Кра...  07.05.2019 09:50   
freq                                                    4                25   

              Авто  
count        44744  
unique           2  
top     Не указано  
freq         32268

----

###  Задание 3.1 ###

Сколько соискателей имеет средний уровень образования (школьное образование)?

***Ответ - 559***

In [8]:
display(main_df['Образование и ВУЗ'])

def get_graduation(string_: str) -> str:
    list_ = string_.split(' ')
    if list_[1] == 'образование':
        return list_[0]
    else:
        return list_[0] + ' ' + list_[1]
    return None

main_df['Образование'] = main_df['Образование и ВУЗ'].apply(get_graduation)
print('Уникальных значений образования -', main_df['Образование'].nunique())
display(main_df['Образование'].value_counts())

0        Неоконченное высшее образование 2000  Балтийск...
1        Высшее образование 1981  Военно-космическая ак...
2        Среднее специальное образование 2002 Профессио...
3        Высшее образование 2002  Саратовский государст...
4        Высшее образование 2015 Кгу Психологии и педаг...
                               ...                        
44739    Высшее образование 2015  Московский гуманитарн...
44740    Высшее образование (Бакалавр) 2016  Воронежски...
44741    Высшее образование 2000  Южно-Уральский госуда...
44742    Высшее образование 2015  Тамбовский государств...
44743    Высшее образование 1997  Южно-Российский госуд...
Name: Образование и ВУЗ, Length: 44744, dtype: object

Уникальных значений образования - 4


Высшее                 33863
Среднее специальное     5765
Неоконченное высшее     4557
Среднее                  559
Name: Образование, dtype: int64

In [9]:
def get_graduation_year(string_: str) -> int:
    return int(re.search(r'\d{4}', string_).group(0))

main_df['Год выпуска'] = main_df['Образование и ВУЗ'].apply(get_graduation_year)
print('Уникальных значений года выпуска -', main_df['Год выпуска'].nunique())
display(main_df['Год выпуска'].value_counts())
display(main_df.info())

Уникальных значений года выпуска - 62


2015    3004
2013    2756
2014    2705
2012    2674
2011    2654
        ... 
2028       3
2029       2
1954       1
1969       1
1972       1
Name: Год выпуска, Length: 62, dtype: int64

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 14 columns):
 #   Column                           Non-Null Count  Dtype 
---  ------                           --------------  ----- 
 0   Пол, возраст                     44744 non-null  object
 1   ЗП                               44744 non-null  object
 2   Ищет работу на должность:        44744 non-null  object
 3   Город, переезд, командировки     44744 non-null  object
 4   Занятость                        44744 non-null  object
 5   График                           44744 non-null  object
 6   Опыт работы                      44576 non-null  object
 7   Последнее/нынешнее место работы  44743 non-null  object
 8   Последняя/нынешняя должность     44742 non-null  object
 9   Образование и ВУЗ                44744 non-null  object
 10  Обновление резюме                44744 non-null  object
 11  Авто                             44744 non-null  object
 12  Образование                     

None

----

### Задание 3.2 ###

Сколько процентов женских резюме представлено в наших данных?

(Ответ округлить до сотых)

***Ответ - 19.07***

Чему равен средний возраст соискателей? Ответ округлите до десятых.

***Ответ - 32.2***

In [10]:
display(main_df['Пол, возраст'])

main_df['Пол'] = main_df['Пол, возраст'].apply(lambda x: x[0])
display(main_df['Пол'].value_counts())
print(round(main_df[main_df['Пол'] == 'Ж'].shape[0] / main_df['Пол'].shape[0] * 100, 2))

0          Мужчина ,  39 лет , родился 27 ноября 1979
1           Мужчина ,  60 лет , родился 20 марта 1959
2        Женщина ,  36 лет , родилась 12 августа 1982
3            Мужчина ,  38 лет , родился 25 июня 1980
4           Женщина ,  26 лет , родилась 3 марта 1993
                             ...                     
44739      Мужчина ,  30 лет , родился 17 января 1989
44740        Мужчина ,  27 лет , родился 5 марта 1992
44741    Женщина ,  48 лет , родилась 26 декабря 1970
44742     Мужчина ,  24 года , родился 6 октября 1994
44743      Мужчина ,  38 лет , родился 25 апреля 1980
Name: Пол, возраст, Length: 44744, dtype: object

М    36211
Ж     8533
Name: Пол, dtype: int64

19.07


In [11]:
def get_age(string_: str) -> int:
    age_str = string_.split(',')[1].strip()
    return int(re.search(r'\d+', age_str).group(0))

##### main_df['Возраст'] = main_df['Пол, возраст'].apply(lambda x: x.split(',')[1].strip())
main_df['Возраст'] = main_df['Пол, возраст'].apply(get_age)
print(round(main_df['Возраст'].mean(), 2))

32.2


----

### Задание 3.3 ###

Чему равен медианный опыт работы (в месяцах) в таблице?

***Ответ - 100***

In [12]:
wtf_test = main_df['Опыт работы'].apply(lambda x: x[:10] if type(x) == str else None)
display(wtf_test.value_counts())

Опыт работ    44574
Не указано        2
Name: Опыт работы, dtype: int64

In [13]:
def convert_experience_to_months(string_: str) -> int:
    try:
        experience = string_.strip('Опыт работы ').split(maxsplit=3)
        if experience[1][0] == 'м':
            experience_months = int(experience[0])
        else:
            experience_months = int(experience[0]) * 12
            if experience[2].isdecimal():
                experience_months += int(experience[2])
        return experience_months
    except:
        return None

main_df['Опыт работы (месяц)'] = main_df['Опыт работы'].apply(convert_experience_to_months)
main_df.drop('Опыт работы', axis='columns', inplace=True)
print('Медианный опыт работы:', main_df['Опыт работы (месяц)'].median())
display(main_df.info())

Медианный опыт работы: 100.0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44744 entries, 0 to 44743
Data columns (total 16 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Пол, возраст                     44744 non-null  object 
 1   ЗП                               44744 non-null  object 
 2   Ищет работу на должность:        44744 non-null  object 
 3   Город, переезд, командировки     44744 non-null  object 
 4   Занятость                        44744 non-null  object 
 5   График                           44744 non-null  object 
 6   Последнее/нынешнее место работы  44743 non-null  object 
 7   Последняя/нынешняя должность     44742 non-null  object 
 8   Образование и ВУЗ                44744 non-null  object 
 9   Обновление резюме                44744 non-null  object 
 10  Авто                             44744 non-null  object 
 11  Образование                      44744 non-null  ob

None

----

### Задание 3.4 ###

Сколько процентов соискателей живут в Санкт-Петербурге?

(Ответ округлить до целого)

***Ответ - 11***

Сколько процентов соискателей готовы одновременно и к переездам, и к
командировкам?

(Ответ округлить до целого)

***Ответ - 32***

In [14]:
def extract_city(string_: str) -> str:
    million_cities = ['Новосибирск',
                      'Екатеринбург',
                      'Нижний Новгород',
                      'Казань',
                      'Челябинск',
                      'Омск',
                      'Самара',
                      'Ростов-на-Дону',
                      'Уфа',
                      'Красноярск',
                      'Пермь',
                      'Воронеж',
                      'Волгоград']
    city = string_.split()[0]
    if city in ['Москва', 'Санкт-Петербург']:
        city_category = city
    elif city in million_cities:
        city_category = 'город-миллионник'
    else:
        city_category = 'другие'
    return city_category

main_df['Город'] = main_df['Город, переезд, командировки'].apply(extract_city)
main_df['Город'] = main_df['Город'].astype('category')
display(main_df['Город'].value_counts())
pitermen = main_df[main_df['Город'] == 'Санкт-Петербург'].shape[0]
print('Процент питерских:', round(pitermen / main_df.shape[0] * 100))

Москва              16621
другие              16452
город-миллионник     6734
Санкт-Петербург      4937
Name: Город, dtype: int64

Процент питерских: 11


In [15]:
print(main_df['Город, переезд, командировки'].values)

['Советск (Калининградская область) , не готов к переезду , не готов к командировкам'
 'Королев , не готов к переезду , готов к редким командировкам'
 'Тверь , не готова к переезду , не готова к командировкам' ...
 'Челябинск , готова к переезду , готова к редким командировкам'
 'Тамбов , не готов к переезду , не готов к командировкам'
 'Москва , не готов к переезду , не готов к командировкам']


In [16]:
def is_ready_for_move(string_: str) -> bool:
    if 'не готов к переезду' in string_ \
        or 'не готова к переезду' in string_:
            return False
    elif 'перее' not in string_:
            return False
    else:
            return True

def is_ready_for_mission(string_: str) -> bool:
    if 'не готов к командировкам' in string_ \
        or 'не готова к командировкам' in string_:
            return False
    elif 'командиров' not in string_:
            return False
    else:
            return True

main_df['Готовность к переезду'] \
    = main_df['Город, переезд, командировки'].apply(is_ready_for_move)
main_df['Готовность к командировкам'] \
    = main_df['Город, переезд, командировки'].apply(is_ready_for_mission)
mask1 = main_df['Готовность к переезду'] == True
mask2 = main_df['Готовность к командировкам'] == True
print('Процент готовых к переезду и командировкам:',
      round(main_df[mask1 & mask2].shape[0] / main_df.shape[0] * 100))

Процент готовых к переезду и командировкам: 32


----
### Задание 3.5 ###

Сколько людей ищут проектную работу и волонтёрство (в обоих столбцах стоит
`True`)?

***Ответ - 436***

Сколько людей хотят работать вахтовым методом и с гибким графиком (в обоих
столбцах стоит `True`)?

***Ответ - 2311***

In [17]:
# hh_set_ini['Занятость'].value_counts()
employments_list = ['полная занятость',
                    'частичная занятость',
                    'проектная работа',
                    'стажировка',
                    'волонтерство'
                   ]

# hh_set_ini['График'].value_counts()
schedules_list = ['полный день',
                  'гибкий график',
                  'сменный график',
                  'вахтовый метод',
                  'удаленная работа'
                 ]

for employment in employments_list:
    main_df[employment] = main_df['Занятость'].aggregate(lambda x: True if employment in x else False, axis='index')

for schedule in schedules_list:
    main_df[schedule] = main_df['График'].aggregate(lambda x: True if schedule in x else False, axis='index')

print('Первый ответ:',
     main_df[main_df['проектная работа'] & main_df['волонтерство']].shape[0])
print('Второй ответ:',
     main_df[main_df['вахтовый метод'] & main_df['гибкий график']].shape[0])

main_df.drop(['Занятость', 'График'], axis='columns', inplace=True)

Первый ответ: 436
Второй ответ: 2311


----
### Задание 3.6 ###

Для выполнения этого задания у вас будет 5 попыток. После того как вы дадите
правильный ответ ИЛИ истратите все попытки, вы можете нажать на кнопку «Показать
ответ»: в конце решения вы увидите ссылку на эталонный ответ.

Чему равна желаемая медианная заработная плата соискателей в нашей таблице (в
рублях)?

Ответ приведите в тысячах рублей и округлите до целого. Например, если у вас
получилась медианная з/п 42 127, в качестве ответа укажите 42.

In [20]:
currencies_df = pd.read_csv('data/ExchangeRates.csv', sep=',')
display(currencies_df['currency'].value_counts())
ini_ser = main_df['ЗП'].apply(lambda x: x.split()[1])
display(ini_ser.value_counts())
tmp_ser = main_df['ЗП'].apply(lambda x: x.split()[1].replace('бел.руб.', 'BYN').replace('грн.', 'UAH').replace('сум', 'UZS'))
display(tmp_ser.value_counts())
main_df['ЗП'] = main_df['ЗП'].apply(lambda x: x.replace('бел.руб.', 'BYN').replace('грн.', 'UAH').replace('сум', 'UZS'))

USD    708
EUR    708
KZT    708
KGS    708
BYN    708
UAH    708
UZS    708
AZN    708
Name: currency, dtype: int64

руб.    42471
KZT      1108
USD       628
BYN       329
EUR       106
UAH        73
UZS        20
KGS         6
AZN         3
Name: ЗП, dtype: int64

руб.    42471
KZT      1108
USD       628
BYN       329
EUR       106
UAH        73
UZS        20
KGS         6
AZN         3
Name: ЗП, dtype: int64

In [29]:
display(main_df['ЗП'].value_counts())

50000 руб.    4064
40000 руб.    3762
30000 руб.    3264
60000 руб.    3087
70000 руб.    2388
              ... 
14525 руб.       1
14900 руб.       1
36500 руб.       1
4860 USD         1
259999 KZT       1
Name: ЗП, Length: 690, dtype: int64

In [27]:
display(main_df['Обновление резюме'])
dt_ser = pd.to_datetime(main_df['Обновление резюме'], format='%d.%m.%Y %H:%M')
display(dt_ser)

0        16.04.2019 15:59
1        12.04.2019 08:42
2        16.04.2019 08:35
3        08.04.2019 14:23
4        22.04.2019 10:32
               ...       
44739    22.04.2019 12:32
44740    22.04.2019 13:11
44741    09.04.2019 05:07
44742    26.04.2019 14:25
44743    05.07.2018 20:15
Name: Обновление резюме, Length: 44744, dtype: object

0       2019-04-16 15:59:00
1       2019-04-12 08:42:00
2       2019-04-16 08:35:00
3       2019-04-08 14:23:00
4       2019-04-22 10:32:00
                ...        
44739   2019-04-22 12:32:00
44740   2019-04-22 13:11:00
44741   2019-04-09 05:07:00
44742   2019-04-26 14:25:00
44743   2018-07-05 20:15:00
Name: Обновление резюме, Length: 44744, dtype: datetime64[ns]

----

### Задание 4.1 ###

Чему равно модальное значение возраста соискателей?

----

### Задание 4.2 ###

Чему равен максимальный опыт работы (в месяцах)?

----

### Задание 4.3 ###

Определите по графику, сколько соискателей требуют заработную плату выше 1 миллиона рублей.

----

###  Задание 4.4 ###

Какая категория образования оплачивается выше всех по медианному показателю?

----

### Задание 4.5 ###

В каком городе (категории городов) зафиксирован наибольший показатель желаемой заработной платы (~924 тысячи рублей)?

----

### Задание 4.6 ###

Чему равна желаемая медианная заработная плата соискателей, готовых и к переезду, и к командировкам? Ответ приведите в тысячах, округлив до целого (например, 45).

----

### Задание 4.7 ###

Для какой категории образования наблюдается самый быстрый карьерный рост (то есть интенсивность роста заработной платы наибольшая)?

----

### Задание 4.8 ###

Сколько точек лежат строго выше построенной прямой? Так как точки могут сливаться с прямой, то для ответа на этот вопрос проще всего воспользоваться интерактивной визуализацией либо получить таблицу с наблюдениями, лежащими выше прямой с помощью фильтрации.